Migrate to matrix-nio

doubleloop opened the pull request:
Fixes #542

thp reviewed (changes requested):
This probably needs a section in the "Deprecated features" documentation on how to migrate from an old matrix configuration to the new one, otherwise people will have their matrix notifications broken (either silently or loudly), which isn't nice. Is there a way we can auto-migrate from the old way to the new way, so Matrix notifications just keep running? Alternatively, should this be a "matrix-nio" reporter and we leave support for the old one in place, but the documentation will just mention that a matrix reporter existed, but new installations should use "matrix-nio"?

thp commented:
Any news on this? Do you think you want to work on getting this ready to be merged?

doubleloop commented:
>needs a section in the "Deprecated features" documentation on how to migrate from an old matrix configuration to the new one, otherwise people will have their matrix notifications broken (either silently or loudly), which isn't nice.

I added special case so that if configuration compatible with `matrix_client` is found, it will still work as expected (with no encryption support). I also added `Deprecated Features` section describing changes required for encryption support.

>Do you think you want to work on getting this ready to be merged?
Yes, waiting for some review/feedback and list of things needed to make this ready to be merged.

thp commented on docs/source/deprecated.rst:
```diff
@@ -5,6 +5,48 @@ This page lists the features that are deprecated and steps to
+----------------------------------------------
+In older versions ``matrix_client`` was used as matrix provider.
+This optional dependency is now replaced with ``matrix-nio`` which is both
+more featureful (support encrypted rooms) and more actively maintained.
```
"support encrypted rooms" -> "supports encrypted rooms"

thp commented on docs/source/deprecated.rst:
```diff
@@ -5,6 +5,48 @@ This page lists the features that are deprecated and steps to
+      user: @botusername:matrix.org
+      enabled: true
+
+Then you need to manually trigger connection to obtain credentials and
+generate encryption keys:
+
+.. code:: bash
+
+    $ urlwatch --test-reporter matrix
+
+This is one time process during which you will be prompted for a password
+to the bot account (password will not be stored).
+Obtained credentials and encryption keys will be stored in user data directory
+approprite for your system (e.g. ``~/.local/share/urlwatch/`` on linux).
+
+Note that it is now not recommended to store ``access_token`` in configuration
+file and it will be only used if ``urlwatch`` recognizes configuration
+compatible with ``matrix_client`` (missing ``user`` field and containing
+``access_token`` field).
```
Just replace this with a link to the Matrix reporter documentation (merging anything that's here and not there to there).

doubleloop replied:
could you verify if I have done it correctly?

thp commented on docs/source/reporters.rst:
```diff
@@ -231,33 +231,37 @@ You can have notifications sent to you through the `Matrix protocol`_.
-   from the bottom.
-8. Close the private browsing window **but do not log out, as this
-   invalidates the Access Token**.
+You then need to acquire room ID, using the following instructions:
```
"to acquire room ID" -> "to acquire the room ID"

thp commented on docs/source/reporters.rst:
```diff
@@ -231,33 +231,37 @@ You can have notifications sent to you through the `Matrix protocol`_.
+This is one time process during which you will be prompted for a password
+to the bot account (password will not be stored).
+Obtained credentials and encryption keys will be stored in user data directory
+approprite for your system (e.g. ``~/.local/share/urlwatch/`` on linux).
```
"approprite" -> "appropriate"

thp replied:
Remove "(e.g. ``~/.local/share/urlwatch/`` on linux)" and replace with something like "(relative to urlwatch's configuration directory)".

thp commented on lib/urlwatch/tests/test_handler.py:
```diff
@@ -90,7 +90,7 @@ def test_pep8_conformance():
     def __init__(self, config, urls, cache, hooks, verbose):
         (prefix, bindir) = os.path.split(os.path.dirname(os.path.abspath(sys.argv[0])))
-        super().__init__('urlwatch', os.path.dirname(__file__), config, urls, cache, hooks, verbose)
+        super().__init__('urlwatch', os.path.dirname(__file__), os.path.dirname(__file__), config, urls, cache, hooks, verbose)
```
Should this use a temporary folder instead of the path where test_handler is? If it's installed, it might not be writable to a user wanting to run the tests.

doubleloop replied:
Made temporary folder. Not sure why `os.path.dirname(__file__)` is used here.
note: what do you thing about renaming `urlwatch_dir` to `urlwatch_config_dir` ?

thp commented on docs/source/deprecated.rst:
```diff
@@ -5,6 +5,48 @@ This page lists the features that are deprecated and steps to
+
+To use encryption you need to update configuration in following ways:
+
+#. Remove ``access_token`` field
```
Should we warn when loading a config that still has `access_token` set?

doubleloop replied:
https://github.com/thp/urlwatch/pull/620/files#diff-5ee1abebe5480db745909e862aaa226d8f66babc4ab506b526c49cf7c0059fe0R1034-R1038

thp commented on lib/urlwatch/reporters.py:
```diff
@@ -935,44 +938,129 @@ def _format_output(self, job_state):
         else:
-            logger.debug('Not formatting as Markdown; dependency on markdown2 not met?')
-            client_api.send_message(room_id, body_markdown)
+            logger.debug(
+                'Not formatting as Markdown; dependency on markdown2 not met?')
```
I know this log message has been like this before, but since we touch it, maybe we can just make it something like:

    markdown2 could not be imported, skipping Markdown formatting

thp commented on lib/urlwatch/handler.py:
```diff
@@ -174,8 +174,9 @@ def _generate_diff(self):
 class Report(object):
-    def __init__(self, urlwatch_config):
-        self.config = urlwatch_config.config_storage.config
+    def __init__(self, urlwatch_obj):
+        self.config = urlwatch_obj.config_storage.config
+        self.urlwatch_config = urlwatch_obj.urlwatch_config
```
Is this change necessary?

thp replied:
Or in other words, could we just set the data directory as additional parameter, not expose the whole config object?

doubleloop replied:
Not sure why not to expose the whole config (other features might require access to command line args or other paths like cache), but I will change this.

> could we just set the data directory as additional parameter
Note that I only renamed existing parameter and did not introduce any new one (`urlwatch_config` -> `urlwatch_obj`). Previous name was misleading (https://github.com/thp/urlwatch/blob/master/lib/urlwatch/main.py#L54)

thp reviewed (changes requested):
See comments. We're getting there :)

doubleloop commented:
I rebased on master and added `review fix` commits (I plan to squash them before merge).

thp commented on lib/urlwatch/reporters.py:
```diff
@@ -935,44 +938,129 @@ def _format_output(self, job_state):
+                self.load_store()
+
+            if not self.user_id or not self.device_id or not self.access_token:
+                logger.debug("Generating new credentials")
+                print("### Generating new Matrix credentials. ###")
+                device_name = input(f"Choose a name for this device: [{self.DEFAULT_DEVICE_NAME}] ")
+                if not device_name:
+                    device_name = self.DEFAULT_DEVICE_NAME
+                pw = getpass.getpass(f"Matrix password for {self.user} (it will not be stored): ")
```
This should most likely be a separate mode from the command line, as we don't want random invocations to be interactive (urlwatch is most likely running as a cronjob or via some other non-interactive means).

thp replied:
See the `--xmpp-login` command-line option for an example. Maybe combine it like `--test-filter` so that you can do e.g. `--login-reporter xmpp` and `--login-reporter matrix` (as `--xmpp-login` is an interactive thing, maybe just update docs + mention it in the list of deprecated things that `--xmpp-login is now `--login-reporter xmpp`).

thp commented on lib/urlwatch/reporters.py:
```diff
@@ -935,44 +938,129 @@ def _format_output(self, job_state):
+                    logger.debug(f'joined room {room_id}')
+                if client.should_upload_keys:
+                    await client.keys_upload()
+                await client.sync(timeout=30000, full_state=True)
```
`30000` should probably be an named constant (`MATRIX_CLIENT_SYNC_TIMEOUT_MS`?)

thp commented on docs/source/reporters.rst:
```diff
@@ -231,33 +232,37 @@ You can have notifications sent to you through the `Matrix protocol`_.
-   invalidates the Access Token**.
+You then need to acquire the room ID, using the following instructions:
+
+#. Open `Riot.im <https://riot.im/app/>`__ in a private browsing window
```
Is there some command-line matrix client that can be used for this without relying on `riot.im` or a similar service? I know that this was in here before, but since one of the advantages of this change is that rooms are now encrypted, the encryption would be kinda pointless if initially we rely on trusting `riot.im`.

doubleloop replied:
> Is there some command-line matrix client that can be used for this without relying on riot.im or a similar service?

According to spec, room `name` is not unique, room `alias` is optional so `room ID` in general is the only reliable room identifier (https://matrix.org/docs/spec/client_server/latest#m-room-name). `room ID` is obtained during room creation and stored by the client. Any client used to create room should be able to display `room ID` (I can note in docs that ~riot.im~ element is just an example)

>the encryption would be kinda pointless if initially we rely on trusting riot.im

Why?

thp reviewed (changes requested):
See comments. We're getting there!

thp commented:
@doubleloop Any updates on this PR?

thp commented:
Closing for now. Please reopen once rebased.
